File: gokdogan-onboard/src/gokdogan_common/test/test_qos.cpp
// QoS profil tutarlılık testi (C++). SAD §6 golden değerleri — qos.py ile PARİTE kontratı.
// Aynı golden test_qos.py'de; ikisi eşleşmezse C++/Python node'ları sessizce kopar (prompt §2.5).
#include <gtest/gtest.h>
#include "gokdogan_common/qos.hpp"

namespace
{
struct Golden
{
  rmw_qos_reliability_policy_t rel;
  rmw_qos_durability_policy_t dur;
  size_t depth;
};

void check(const rclcpp::QoS & q, const Golden & g, const char * name)
{
  const auto & p = q.get_rmw_qos_profile();
  EXPECT_EQ(p.history, RMW_QOS_POLICY_HISTORY_KEEP_LAST) << name;
  EXPECT_EQ(p.reliability, g.rel) << name;
  EXPECT_EQ(p.durability, g.dur) << name;
  EXPECT_EQ(p.depth, g.depth) << name;
}
}  // namespace

TEST(QoSParity, MatchesSadGolden)
{
  using namespace gokdogan_common;
  check(sensor_stream(),
        {RMW_QOS_POLICY_RELIABILITY_BEST_EFFORT, RMW_QOS_POLICY_DURABILITY_VOLATILE, 1},
        "sensor_stream");
  check(detections(),
        {RMW_QOS_POLICY_RELIABILITY_RELIABLE, RMW_QOS_POLICY_DURABILITY_VOLATILE, 5},
        "detections");
  check(target_selected(),
        {RMW_QOS_POLICY_RELIABILITY_RELIABLE, RMW_QOS_POLICY_DURABILITY_TRANSIENT_LOCAL, 10},
        "target_selected");
  check(lock_event(),
        {RMW_QOS_POLICY_RELIABILITY_RELIABLE, RMW_QOS_POLICY_DURABILITY_VOLATILE, 20},
        "lock_event");
  check(mission_mode(),
        {RMW_QOS_POLICY_RELIABILITY_RELIABLE, RMW_QOS_POLICY_DURABILITY_TRANSIENT_LOCAL, 10},
        "mission_mode");
  check(mission_command(),
        {RMW_QOS_POLICY_RELIABILITY_RELIABLE, RMW_QOS_POLICY_DURABILITY_VOLATILE, 10},
        "mission_command");
  check(server_data(),
        {RMW_QOS_POLICY_RELIABILITY_RELIABLE, RMW_QOS_POLICY_DURABILITY_TRANSIENT_LOCAL, 5},
        "server_data");
}

File: gokdogan-onboard/src/gokdogan_common/test/test_qos.py
"""QoS profil tutarlılık testi (SAD §6). Golden değerler bağımsız olarak burada tanımlı;
qos.py sapmaları yakalanır. Bu ayrıca C++ (qos.hpp) parite kontratıdır — aynı golden
değerler test_qos.cpp'de doğrulanır (pub/sub sessiz kopmasını önler, prompt §2.5)."""
from rclpy.qos import (
    QoSReliabilityPolicy,
    QoSDurabilityPolicy,
    QoSHistoryPolicy,
)

from gokdogan_common import qos

R_BE = QoSReliabilityPolicy.BEST_EFFORT
R_R = QoSReliabilityPolicy.RELIABLE
D_V = QoSDurabilityPolicy.VOLATILE
D_T = QoSDurabilityPolicy.TRANSIENT_LOCAL
H_KL = QoSHistoryPolicy.KEEP_LAST

# SAD §6 QoS matrisi — (reliability, durability, depth)
GOLDEN = {
    "sensor_stream":   (R_BE, D_V, 1),
    "detections":      (R_R,  D_V, 5),
    "target_selected": (R_R,  D_T, 10),
    "lock_event":      (R_R,  D_V, 20),
    "mission_mode":    (R_R,  D_T, 10),
    "mission_command": (R_R,  D_V, 10),
    "server_data":     (R_R,  D_T, 5),
}


def test_all_profiles_present():
    assert set(qos.PROFILES.keys()) == set(GOLDEN.keys())


def test_profiles_match_sad_golden():
    for name, (rel, dur, depth) in GOLDEN.items():
        p = qos.PROFILES[name]()
        assert p.history == H_KL, f"{name}: history KEEP_LAST değil"
        assert p.reliability == rel, f"{name}: reliability {p.reliability} != {rel}"
        assert p.durability == dur, f"{name}: durability {p.durability} != {dur}"
        assert p.depth == depth, f"{name}: depth {p.depth} != {depth}"


def test_factories_return_fresh_instances():
    # Her çağrı bağımsız profil döndürmeli (paylaşılan mutable state yok)
    a = qos.sensor_stream()
    b = qos.sensor_stream()
    assert a is not b
